web-common/src/features/workspaces/CanvasWorkspace.svelte

```diff
@@ -0,0 +1,196 @@
+  }
+
+  async function addComponent(componentName: CanvasComponentType) {
+    const defaultMetrics = $metricsViewQuery?.data;
+    if (!defaultMetrics) return;
+
+    const newSpec = componentRegistry[componentName].newComponentSpec(
+      defaultMetrics.metricsView,
+      defaultMetrics.measure,
+      defaultMetrics.dimension,
+    );
+
+    const { width, height } = componentRegistry[componentName].defaultSize;
+    const newComponent = {
+      component: { [componentName]: newSpec },
+      height,
+      width,
+      x: 0,
+      y: 0,
+    };
+    const parsedDocument = parseDocument(
+      $editorContent ?? $remoteContent ?? "",
+    );
+
+    const items = parsedDocument.get("items") as any;
+
+    if (!items) {
+      parsedDocument.set("items", [newComponent]);
+    } else {
+      items.add(newComponent);
+    }
+
+    updateEditorContent(parsedDocument.toString(), true);
+    await updateComponentFile();
+  }
+</script>
+
+{#if canvasResource && fileArtifact}
+  {#key canvasName}
+    <StateManagersProvider {canvasName}>
+      <CanvasThemeProvider>
+        <WorkspaceContainer>
+          <WorkspaceHeader
+            slot="header"
+            {filePath}
+            hasUnsavedChanges={$hasUnsavedChanges}
+            titleInput={fileName}
+            onTitleChange={onChangeCallback}
+            resourceKind={ResourceKind.Canvas}
+          >
+            <div class="flex gap-x-2" slot="cta">
+              <PreviewButton
+                href="/canvas/{canvasName}"
+                disabled={allErrors.length > 0 || resourceIsReconciling}
+                reconciling={resourceIsReconciling}
+              />
+
+              <AddComponentMenu {addComponent} />
+              <ViewSelector
+                allowSplit={false}
+                bind:selectedView={$selectedViewStore}
+              />
+            </div>
+          </WorkspaceHeader>
+
+          <WorkspaceEditorContainer
+            slot="body"
+            error={mainError}
+            showError={!!$remoteContent && selectedView === "code"}
+          >
+            {#if selectedView === "code"}
+              <CanvasEditor
+                bind:autoSave={$autoSave}
+                {canvasName}
+                {fileArtifact}
+                {lineBasedRuntimeErrors}
+              />
+            {:else if selectedView === "viz"}
+              {#if mainError}
+                <ErrorPage
+                  body={mainError.message}
+                  fatal
+                  header="Unable to load dashboard preview"
+                  statusCode={404}
+                />
+              {:else if canvasResource}
+                <Canvas {fileArtifact} />
+              {/if}
+            {/if}
+          </WorkspaceEditorContainer>
+
+          <VisualCanvasEditing {fileArtifact} slot="inspector" />
+        </WorkspaceContainer>
+      </CanvasThemeProvider>
+    </StateManagersProvider>
+  {/key}
+{:else if allErrors.length}
```

Comment:
This should not be in the else. Since both `code` and `viz` are in the other if body, if there are errors user cannot edit the canvas in rill-dev at all.